picard/extension_points/cover_art_processors.py

```diff
@@ -0,0 +1,65 @@
+# -*- coding: utf-8 -*-
+#
+# Picard, the next-generation MusicBrainz tagger
+#
+# Copyright (C) 2024 Giorgio Fontanive
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+
+from enum import IntEnum
+
+from picard.plugin import ExtensionPoint
+
+
+ext_point_cover_art_processors = ExtensionPoint(label='cover_art_processors')
+
+
+class ProcessingTarget(IntEnum):
+    TAGS = 0
+    FILE = 1
+    BOTH = 2
+
+
+class ImageProcessor:
+
+    def save_to_tags(self):
+        return False
+
+    def save_to_file(self):
+        return False
+
+    def same_processing(self):
+        return False
+
+    def run(self, data, info, target):
+        pass
+
+
+def get_cover_art_processors():
+    queue_both, queue_tags, queue_file = [], [], []
+    for processor in ext_point_cover_art_processors:
+        if processor.same_processing():
+            queue_both.append(processor)
+        else:
+            if processor.save_to_tags():
+                queue_tags.append(processor)
+            if processor.save_to_file():
+                queue_file.append(processor)
+    return queue_both, queue_tags, queue_file
```

Comment:
We could perhaps make this a generator and yield `ProcessingTarget` along:

```python
def get_cover_art_processors():
    for processor in ext_point_cover_art_processors:
        if processor.same_processing():
            yield (processor, ProcessingTarget.BOTH)
        else:
            if processor.save_to_tags():
                yield (processor, ProcessingTarget.TAGS)
            if processor.save_to_file():
                yield (processor, ProcessingTarget.FILE)
  ```

And in the code calling this method, something like:

```python
        processed_data = dict()
        for processor, target in get_cover_art_processors():
            processed_data[target] = processor.run(data, info, target)
```

What do you think?

Reply:
The issue is that we need to execute all processors with `same_processing = True` first. Then once all of those are finished we run the two other target specific queues with the data obtained from the first. I think using a generator like this does not ensure the order of the different queues, or am I wrong? 

Once I start working on adding multi-threading, I will have to change the `run_image_processors` function anyways, because I was planning on having the processing chain for each target run in a separate thread.

Reply:
> The issue is that we need to execute all processors with `same_processing = True` first. Then once all of those are finished we run the two other target specific queues with the data obtained from the first. I think using a generator like this does not ensure the order of the different queues, or am I wrong?

Hmmm, right.
We can still use a generator, but explode the loops:

```python
def get_cover_art_processors():
    not_same_processing = []
    for processor in ext_point_cover_art_processors:
        if processor.same_processing():
            yield (processor, ProcessingTarget.BOTH)
        else:
            not_same_processing.append(processor)
    for processor in not_same_processing:
        if processor.save_to_tags():
            yield (processor, ProcessingTarget.TAGS)
        if processor.save_to_file():
            yield (processor, ProcessingTarget.FILE)
```